### ISSUE TITLE
test_material.py - UnicodeDecodeError

### ISSUE THREAD

**Issue by sturmk** [pyne/pyne#1535]
**Describe the Bug**<br/>
When using any functionality related to materials, I get an UnicodeDecodeError.
Specifically, running the test `test_material.py` triggers this issue reliably.

**To Reproduce**<br/>
Run pyne tests.

**Expected Behavior**<br/>
No error.

**Screenshots or Code Snippets**<br/>
```
===================================================================== test session starts =====================================================================
platform linux -- Python 3.11.6, pytest-7.4.0, pluggy-1.2.0
rootdir: /home/karl/tools/pyne
collected 108 items                                                                                                                                           

test_material.py ............................................................F...............................................                           [100%]

========================================================================== FAILURES ===========================================================================
______________________________________________________________________ test_getitem_str _______________________________________________________________________

    def test_getitem_str():
        mat = Material(nucvec)
        assert mat["U235"] == 1.0
>       pytest.raises(RuntimeError, lambda: mat["word"])

test_material.py:769: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_material.py:769: in <lambda>
    pytest.raises(RuntimeError, lambda: mat["word"])
material.pyx:1485: in pyne.material._Material.__getitem__
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb2 in position 1: invalid start byte

nucname.pyx:188: UnicodeDecodeError
=================================================================== short test summary info ===================================================================
FAILED test_material.py::test_getitem_str - UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb2 in position 1: invalid start byte
================================================================ 1 failed, 107 passed in 0.47s ================================================================
Testing failed on test_material.py
```

**Please complete the following information regarding your system:**<br/>
 - OS: Ubuntu 23.10
 - PyNE Version: 0.7.8
 - Versions of dependencies installed with: native Ubuntu python dependencies, no venv

**Comment by sturmk** [comment]
Update: seems to be a general issue with Pyne on Ubuntu 23.10.
On Ubuntu 22.04 LTS, installation and usage of pyne is flawless.
This would also be interesting to try on the freshly released Ubuntu 24.04 LTS

**Comment by gonuke** [comment]
We'll definitely have to dig into the 24.04 and see what new surprises await!

**Comment by ahnaf-tahmid-chowdhury** [comment]
Python has been updated, and we need to update our code accordingly. The current logic of the code is that if it sends a RuntimeError, the test will pass. However, with the new update, I think, Python>3.10 also checks for a UnicodeDecodeError. Therefore, we need to modify our code to include this as well.

```
pytest.raises([RuntimeError, UnicodeDecodeError], lambda: mat["word"])
```

**Comment by gonuke** [comment]
I think this may be resolved by #1544?

**Comment by bquan0** [comment]
Hi @sturmk, I think this issue has been fixed in #1543. The `UnicodeDecodeError`s were caused by the `id()` function in `nucname.cpp` whenever they threw a custom exception. These custom exceptions had a `what()` function which would provide an error message in the form of a `const char*` pointer. The problem with the implementation was that the pointer referred to a string created in the `what()` function, which would be deallocated once the function returned. Thus, the pointer would become invalid. When Cython tries to translate this message, it would try to decode the message at the invalid pointer, which would cause the `UnicodeDecodeError`. 

The problem has now been fixed by adding a new string to the custom exception that stores the error message. This way, the string won't be deallocated and the pointer will still be valid.

Let me know if you're still running into this error. Otherwise, the issue can be closed if the error has been resolved.

**Comment by sturmk** [comment]
sorry for the late reply, I only just got around to testing it. the tests now run through without much trouble on Ubuntu 23.10 & Python 3.11, thanks for fixing!
